pkg/ottl/contexts/internal/value.go

```diff
@@ -164,7 +180,20 @@ func setIndexableValue[K any](ctx context.Context, tCtx K, currentValue pcommon.
 				}
 				currentValue = currentValue.Slice().AppendEmpty()
 			default:
-				return errors.New("neither a string nor an int index was given, this is an error in the OTTL")
+				resString, errString := FetchValueFromExpression[K, string](ctx, tCtx, keys[count])
+				resInt, errInt := FetchValueFromExpression[K, int64](ctx, tCtx, keys[count])
```

Comment:
Would it be possible to take these out of this default case and consider all four cases inside a single switch statement?